src/opam_metadata.ml

```diff
@@ -360,9 +360,21 @@ let nix_of_opam ~pkg ~deps ~(opam_src:opam_src) ~opam ~src ~url () : Nix_expr.t
 		|> List.sort (fun (a,_) (b,_) -> String.compare a b)
 	in
 
-	let opam_inputs : Nix_expr.t AttrSet.t =
-		!opam_inputs |> InputMap.mapi (fun name importance ->
-			property_of_input (`Id "selection") (name, importance)) in
+	let opam_inputs : Nix_expr.attrset =
+		let (expr, inherits) =
+		InputMap.fold (fun name importance (exprs, inherits) ->
+			match importance with
+				| Optional -> (`Expr (name, `Property_or (`Id "selection", name, `Null)) :: exprs, inherits)
+				| Required ->
+					begin match String.split_on_char '.' name with
+					| [ name ] -> (exprs, name :: inherits)
+					| names ->
+						`Expr (name, `PropertyPath (`Id "selection", names)) :: exprs,
+						inherits
+					end) !opam_inputs ([], [])
+		in
+		`Inherit (Some (`Id "selection"), inherits) :: expr
+	in
```

Comment:
Can you do this to `property_of_input` instead of inline? That function is used in two places, and the other one (nix depexts) would benefit from the same treatment.

(because of the accumulation, maybe that function would need to cover the whole set of bindings, which is fine)

Reply:
By depexts you mean `nix_deps`, right? But that's a list rather than an attribut set. So I don't see how this would apply.

Reply:
Ahh right, yep. I just saw that they used to share an implementation and now don't, which is a shame.